pallets/claim/src/lib.rs

```diff
@@ -0,0 +1,467 @@
+// Copyright 2024, Horizen Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+#![deny(missing_docs)]
+
+//! A pallet implementing the possibility of making airdrops on-chain and letting **Beneficiaries**
+//! manually claim the amount of tokens they have right to.
+//! Only **ManagerOrigin** is able to start and end airdrops, as well as adding beneficiaries with
+//! their rightful balance.
+//! Currently it possible only to held one airdrop at a time.
+//! When airdrop ends, all the funds still available in the pallet's associated account
+//! are transferred to **UnclaimedDestination**.
+//!
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+mod weight;
+use core::marker::PhantomData;
+
+extern crate alloc;
+
+use alloc::{collections::btree_map::BTreeMap, vec::Vec};
+use sp_runtime::traits::{AccountIdConversion, Saturating, Zero};
+
+use frame_support::{
+    dispatch::{DispatchResult, PostDispatchInfo},
+    traits::{
+        fungible::{Inspect, Mutate},
+        tokens::{Fortitude, Pay, Preservation},
+        Get,
+    },
+    PalletId,
+};
+
+pub(crate) type BalanceOf<T> =
+    <<T as Config>::Currency as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+
+pub use pallet::*;
+pub use weight::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+
+    use super::*;
+    use frame_support::pallet_prelude::*;
+    use frame_system::pallet_prelude::*;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    #[pallet::config]
+    pub trait Config: frame_system::Config {
+        /// The overarching event type.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+        /// The claim's pallet id, used for deriving its sovereign account ID.
+        #[pallet::constant]
+        type PalletId: Get<PalletId>;
+
+        /// Manager allowed to begin/end airdrops and add beneficiaries
+        type ManagerOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+        /// Type for processing spends of [Self::AssetKind] in favor of [`Self::Beneficiary`].
+        type Paymaster: Pay<
+            Beneficiary = Self::AccountId,
+            AssetKind = (),
+            Balance = BalanceOf<Self>,
+        >;
```

Comment:
Maybe you don't need this abstraction at all. Indeed, it's nice, but you cannot use this abstraction to inspect the account and to move founds in. Moreove this can be not coherent with the pallet `account_id` used to check the ballance and move the unclaimed founds.